tensorflow_serving/tools/docker/Dockerfile.mkl

```diff
@@ -0,0 +1,98 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+ARG TF_SERVING_VERSION=latest
+ARG TF_SERVING_BUILD_IMAGE=tensorflow/serving:${TF_SERVING_VERSION}-devel-mkl
+
+FROM ${TF_SERVING_BUILD_IMAGE} as build_image
+FROM ubuntu:18.04
+
+ARG TF_SERVING_VERSION_GIT_BRANCH=master
+ARG TF_SERVING_VERSION_GIT_COMMIT=head
+
+LABEL maintainer="Karthik Vadla <karthik.vadla@intel.com>"
+LABEL tensorflow_serving_github_branchtag=${TF_SERVING_VERSION_GIT_BRANCH}
+LABEL tensorflow_serving_github_commit=${TF_SERVING_VERSION_GIT_COMMIT}
+
+RUN apt-get update && apt-get install -y --no-install-recommends \
+        ca-certificates htop \
+        && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/*
+
+# Install TF Serving pkg
+COPY --from=build_image /usr/local/bin/tensorflow_model_server /usr/bin/tensorflow_model_server
+
+# Install MKL libraries
+COPY --from=build_image /usr/local/lib/libiomp5.so /usr/local/lib
+COPY --from=build_image /usr/local/lib/libmklml_gnu.so /usr/local/lib
+COPY --from=build_image /usr/local/lib/libmklml_intel.so /usr/local/lib
+
+ENV LIBRARY_PATH '/usr/local/lib:$LIBRARY_PATH'
+ENV LD_LIBRARY_PATH '/usr/local/lib:$LD_LIBRARY_PATH'
+
+
+# Expose ports
+# gRPC
+EXPOSE 8500
+
+# REST
+EXPOSE 8501
+
+# Set where models should be stored in the container
+ENV MODEL_BASE_PATH=/models
+RUN mkdir -p ${MODEL_BASE_PATH}
+
+# The only required piece is the model name in order to differentiate endpoints
+ENV MODEL_NAME=model
+
+# Create a script that runs the model server so we can use environment variables
+# while also passing in arguments from the docker command line
+
+# Setting MKL environment variables gave better performance.
+# https://www.tensorflow.org/guide/performance/overview
+# Read about Tuning MKL for the best performance
+# Add export MKLDNN_VERBOSE=1 to the below script,
+# to see MKL messages in the docker logs when you send predict request.
+
+# Based on our observations during experiments,
+# setting tensorflow_session_parallelism=<1/4th of physical cores> and
+# setting OMP_NUM_THREADS=<Total physical cores>
+# gave optimal performance results with MKL
+# KMP_BLOCKTIME=<Varies based on your model>
+
+# NOTE: We don't gurantee same settings to give optimal peformance across all hardwares
+# please tune variables as required.
+
+ENV OMP_NUM_THREADS=2
+ENV KMP_BLOCKTIME=1
+ENV KMP_SETTINGS=1
+ENV KMP_AFFINITY='granularity=fine,verbose,compact,1,0'
+ENV MKLDNN_VERBOSE=0
+
+ENV TENSORFLOW_SESSION_PARALLELISM=2
+
+RUN echo '#!/bin/bash \n\n\
+export OMP_NUM_THREADS=${OMP_NUM_THREADS} \n\
```

Comment:
Are these exports actually necessary? Shouldn't the above declarations force them to be set in the environment?

Reply:
Note that we don't have to export MODEL_NAME or TENSORFLOW_SESSION_PARALLELISM

Reply:
`MODEL_NAME` is same from your default `Dockerfile`. 
`TENSORFLOW_SESSION_PARALLELISM` env is the key for getting optimal performance. During experiments our data scientists suggested to expose that as env so that it easy to set cores. Otherwise default is `2`.

Reply:
> Are these exports actually necessary? Shouldn't the above declarations force them to be set in the environment?

yeah we need these to set to get optimal performance with `mkl`
```
ENV OMP_NUM_THREADS=2
ENV KMP_BLOCKTIME=1
ENV KMP_SETTINGS=1
ENV KMP_AFFINITY='granularity=fine,verbose,compact,1,0'
```

Reply:
but why add "export" to the script?

Reply:
I'm just saying that they're already exported by Docker

Reply:
In other words, shouldn't this just be:

```shell
RUN echo '#!/bin/bash \n\n\
tensorflow_model_server --port=8500 --rest_api_port=8501 \
--tensorflow_session_parallelism=${TENSORFLOW_SESSION_PARALLELISM} \
--model_name=${MODEL_NAME} --model_base_path=${MODEL_BASE_PATH}/${MODEL_NAME} \
"$@"' > /usr/bin/tf_serving_entrypoint.sh \
&& chmod +x /usr/bin/tf_serving_entrypoint.sh
```

Since all the environment variables are already exported (just like MODEL_NAME and TENSORFLOW_SESSION_PARALLELISM)

Reply:
You are right, removed.